### PR TITLE
gatekeeper: persist guardduty findings in big query

### PR DESCRIPTION
**Pull request by adrianosela**
closes https://github.com/mozilla-services/foxsec-pipeline/issues/207,

Updates in terraform: 

- NEW big query dataset for guardduty findings
- NEW big query table for guardduty findings, with schema:
```
[
  {
    "name": "awsAccountId",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "guarddutyFindingId",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "guarddutyFindingType",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "raw",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "timestamp",
    "type": "TIMESTAMP",
    "mode": "NULLABLE"
  }
]
```

Updates in deployment script:

- NEW argument: 
```--findingsPersistenceBigQuery=${GOOGLE_PROJECT_ID}:${PERSIST_DATASET_NAME}.${PERSIST_TABLE_NAME}``` 

Both these things live in https://github.com/mozilla-services/secops-gatekeeper-poc master branch

### NOTES: this is not yet deployed in dataflow, but the terraform has been ran

### CODE REVIEW

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyTransforms.java [hunk 1]
```diff
@@ -122,6 +132,78 @@ public void processElement(ProcessContext c) {
     }
   }
 
+  /** Persist GuardDuty findings onto GCP's Big Query */
+  public static class PersistFindings extends PTransform<PCollection<Event>, PDone> {
+    private static final long serialVersionUID = 1L;
+
+    private static ObjectMapper mapper = new ObjectMapper();
+    private static String findingsPersistenceBigQuery;
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyTransforms.java]: Any reason this needs to be static?

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyTransforms.java [hunk 2]
```diff
@@ -122,6 +132,78 @@ public void processElement(ProcessContext c) {
...
+    private TableRow FindingsTableRow(Finding f) {
+      TableRow r = new TableRow();
+      r.set("awsAccountId", f.getAccountId());
+      r.set("guarddutyFindingId", f.getId());
+      r.set("guarddutyFindingType", f.getType());
+      r.set("timestamp", f.getUpdatedAt());
+
+      String raw;
+      try {
+        raw = mapper.writeValueAsString(f);
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyTransforms.java]: Is this the only way to do this? Is some sort of conversion from the JSON to a row that would create a column for each field not possible?

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyTransforms.java [hunk 3]
```diff
@@ -122,6 +132,78 @@ public void processElement(ProcessContext c) {
...
+     *
+     * @param opts {@link Options} pipeline options
+     */
+    public PersistFindings(Options opts) {
+      findingsPersistenceBigQuery = opts.getFindingsPersistenceBigQuery();
+    }
+
+    @Override
+    public PDone expand(PCollection<Event> input) {
+      if (findingsPersistenceBigQuery != null) {
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/gatekeeper/GuardDutyTransforms.java]: Can we just remove this `if`? This transform should only be initialized when this option is set anyway no?

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyTransforms.java (hunk 1 above)
Comment (adrianosela) [reply src/main/java/com/mozilla/secops/gatekeeper/GuardDutyTransforms.java]: nope 👍

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyTransforms.java (hunk 3 above)
Comment (adrianosela) [reply src/main/java/com/mozilla/secops/gatekeeper/GuardDutyTransforms.java]: yes we can 👍

File: src/main/java/com/mozilla/secops/gatekeeper/GuardDutyTransforms.java (hunk 2 above)
Comment (ajvb) [reply src/main/java/com/mozilla/secops/gatekeeper/GuardDutyTransforms.java]: I'll just call out that this was per my suggestion. My thinking here was that dealing with migrations if/when new GD fields were changed/removed/added might be annoying and we initially just wanted a centralized place to fetch the raw GD finding.

Comment (ameihm0912) [reply src/main/java/com/mozilla/secops/gatekeeper/GuardDutyTransforms.java]: Ok yeah, I wasn't sure if BQ had a nice way to handle arbitrary JSON or something like this, but that makes sense

**Comment by adrianosela** [comment]
> I feel as though this would be nicer if it was implemented at the global level, possibly by extending event payload classes to support something like `toTableRow`, and then just implementing this functionality in `Composite{Input,Output}`.

opened https://github.com/mozilla-services/foxsec-pipeline/issues/229 and closing this PR @ajvb @ameihm0912

**Pull request closed without merging**